### ISSUE TITLE
WPML : Way To Select Language Synced with FB -> Fix Error

### ISSUE THREAD

**Issue by christos312** [facebookincubator/facebook-for-woocommerce#279]
Hello, I installed the plugin and all the products have synced with my FB Page, but half of them are not published.
I get this message in Facebook
> People cannot currently see this product because it is not marked as published for sale on Facebook.

I looked everywhere and I cannot find a way to publish the products.

In Woocomerce products are published.

Is someone having the same issue?

**Comment by jieminz22** [comment]
Hi,

Which version of the extension are you using? Are those products visible on Woocommerce?

**Comment by savva312** [comment]
Hello, I am using Version 1.7.5.

Products are visible on Woocommerce, they are Published and In Stock.

**Comment by christos312** [comment]
Hello, I am using Version 1.7.5.

Products are visible on Woocommerce, they are Published and In Stock.

**Comment by dmitridr** [comment]
Hello,

Could you give us a link to or a screenshot of one of these products? As well as a screenshot of the configuration/ settings page of the product? 

We do set some of the products to hidden, for example products that are in the trash will be marked as hidden and you will see the message you gave.

**Comment by christos312** [comment]
Hi dmitridr

Please see below

Screenshot in facebook (shop page)
[http://prntscr.com/iebu98](url)

Screenshot of product details in facebook
[http://prntscr.com/iebuut](url)

Screenshot from WC
[http://prntscr.com/iebvpz](url)

After some more investigation from my side, I realized that the products that are hidden in FB are the ones I have for the second language of my site.

I am using WPML and I have English and Greek language. The English products are ok in FB, the Greek ones are not published.

Another thing I noticed is this message
[http://prntscr.com/ieby1h](url)

Probably this is due to the different URL for checkout in English and Greek?

In the English version the url is MYDOMAIN/checkout
In Greek is MYDOMAIN/el/checkout

I don't know if these info help.

**Comment by dmitridr** [comment]
Hello,

Thanks for the screenshots and follow up info, I understand the issue now.

The language behavior for WPML is something we added recently... before the behavior was to upload both languages and there were lots of complaints that most people wanted only one language uploaded to FB. So now we hide the products in the shop unless they are in the default language.

Would you prefer to show all these products no matter what the language? I can suggest a workaround for you, if you want to do this. But this will duplicate the products that you have in 2 languages and show both copies on FB.

If not, what would be the behavior you want from the plugin? We can consider adding it as a feature.
Thanks!

P.S. Yes the checkout URL notice is probably due to the the different language. We will look into changing that notice.

**Comment by christos312** [comment]
So this is by design. Ok thats good to know.

Well my initial idea was to create 2 collections on FB shop, one for English and one for Greek.

To do that I would need both languages to be visible on FB.

So if it is possible to select the language I want to sync to FB that would be great.
Something like a setting in WP where it lists all languages I enabled in WPML and I select 1 or more to be synced with FB.

Thanks again for your time! Much appreciated

**Comment by Jekil82** [comment]
I have some problem:
2 lenguages (WPML), some product only in one lenguage (not-default lenguage) and this are hidden on fb page. 

I think the best solution are a check box to choose wath we want show on fb shop page

Thanks for your work!

**Comment by dmitridr** [comment]
Thanks all, it's good to know there is more demand for this feature. Maybe we can hook into the WPML settings page to provide this setting. We will look into possibilities, though it may be a while before we get to this, so (for any devs reading this) would love if anyone wants to send us a pull request!

The workaround to sync both in the meantime is to delete [these lines](https://github.com/facebookincubator/facebook-for-woocommerce/blob/v1.7.11/includes/fbproduct.php#L443-L454) using the plugin editor. Then, if you do a Force Resync, it should unhide all non-default language products in the shop.

@Jekil82 for your second problem, when you click the hide/show button does nothing happen on the FB shop side? When you use the Hide button, the product should display something like `this product is currently not visible or published` when you click on details in the fb shop. (Feel free to open a separate issue on this)

**Comment by aavvaavv** [comment]
Hello. I found the same problem. But also, I think that additional but. Now I will describe and show
1. WPML
2. 2 languages.
3. 304 goods; http://prntscr.com/iie768 - 331 main language and 158 on additional language. 173 of 331 on original language is private goods.

So, when I start to sync, I see next message from plugin: http://prntscr.com/iie8gi (We have 158 goods)
158 on main language and 158 on additional. = 316 goods. 
When I make re sync, it shows 304 - http://prntscr.com/iie90f m I thin few goods are private.

So, after sync, with variations, I see next: http://prntscr.com/iie9ss - 2800 products, => http://prntscr.com/iiea1a - 2752 variations.
And, on the FB show I see next - http://prntscr.com/iieai4 - 124 items when logged, and 124\2=62 from other account - without duplicates - http://prntscr.com/iiecw9

The main idea, to which we will return - I synced 124 goods * = 248(2 langs)(2800 with variations), but on FB I see only 62.
If I syncing 98 goods * 2 = 196 on 2 languages, I see only 196 \ 2 \ 2 = 49 goods on FB.
I think that there is a limit on the display of goods. How many downloaded, so many can be displayed. But because of takes in the second language, I can not show the remaining half of the second goods.

I try to mark of next point on the second language: http://prntscr.com/iiefw6

As for me, there is no problem that the goods are unloaded in the second language. But I can not show all 100% of the goods. For some reason, it displays exactly 50% of all goods.

I can provide all accesses and passwords so that you can get acquainted with the situation.
Thank you so much.

Andrey

**Comment by Jany-M** [comment]
@mengyingdu just checking if there will be soon an option to choose from, in the FB plugin?

It would be great a simple "Choose language to sync products from: " or, sync all languages.
Currently (latest FB plugin and WPML) products sync in all languages, which is confusing for everybody... and forces us to delete manually dozens of products.

Even better would be to able to have a multilang Shop tab in Facebook itself :)

(I left the same comment [here](https://github.com/facebookincubator/facebook-for-woocommerce/issues/144), but just noticed that thread is closed and this one isn't)

**Comment by mengyingdu** [comment]
Hi @Jany-M @christos312 @aavvaavv,

We supported WPML in [v1.8.6](https://github.com/facebookincubator/facebook-for-woocommerce/releases/tag/v1.8.6) by adding a language selector. Please clear your browser cache after upgrade.

Thanks.

**Comment by dmitridr** [comment]
@aavvaavv any more details you can give here? What's your WPML version? I'll try getting that version and testing. That's very strange that you get that because that fbutils code should not be related to WPML.

If you could insert the following code on the line in question [fbutils.php line 141](https://github.com/facebookincubator/facebook-for-woocommerce/blob/v1.8.7/includes/fbutils.php#L140), 

```
if ( is_wp_error( $string ) ) {
      $e = new Exception();
      $backtrace = var_export($e->getTraceAsString(), true);
      WC_Facebookcommerce_Utils::fblog(
        "Trace:\n".$backtrace,
        array(),
        true);
}
```

then try saving again, we should be able to see a stack trace in our own logs and be able to understand/solve the problem. Thanks for working with us on this.

**Comment by dmitridr** [comment]
Oh 1.8.6 had known problems, I would also try on 1.8.7 if you can.

**Comment by mengyingdu** [comment]
Hi @aavvaavv ,

This error was fixed in 1.8.7 and the latest version is [v1.9.0](https://github.com/facebookincubator/facebook-for-woocommerce/releases/tag/v1.9.0) Please upgrade.
